### PR TITLE
 [ArrayManager] Array version of interpolate logic

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Similar to what I did in https://github.com/pandas-dev/pandas/pull/41104 for `fillna` (xref https://github.com/pandas-dev/pandas/issues/39146).

This also improves the performance for some of the fillna/interpolate benchmarks (mostly due to avoiding the overhead of creating a block for each column). For example (from ` frame_methods.Fillna.time_frame_fillna(True, 'bfill', 'datetime64[ns]')`:

```python
N, M = 10000, 100
dtype = "datetime64[ns]"
data = pd.date_range("2011-01-01", freq="H", periods=N)
df = pd.DataFrame({f"col_{i}": data for i in range(M)})
df[::2] = None
df_am = df._as_manager("array").copy()

In [2]: %timeit df_am.fillna(inplace=True, method="bfill")
7.31 ms ± 313 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  # <-- master
3.57 ms ± 370 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  # <-- PR
```

### CODE REVIEW

File: pandas/core/missing.py [hunk 1]
```diff
@@ -973,3 +979,66 @@ def _rolling_window(a: npt.NDArray[np.bool_], window: int) -> npt.NDArray[np.boo
...
+        if downcast is None:
+            arr = soft_convert_objects(arr, datetime=True, numeric=False)
+
+    if downcast:
+        arr = maybe_downcast_to_dtype(arr, downcast)
+    return arr
+
+
+def interpolate_array(
+    arr: np.ndarray | ExtensionArray,
```
Comment (jbrockmendel) [inline pandas/core/missing.py:995]: ArrayLike

File: pandas/core/missing.py [hunk 2]
```diff
@@ -973,3 +979,66 @@ def _rolling_window(a: npt.NDArray[np.bool_], window: int) -> npt.NDArray[np.boo
...
+        axis=axis,
+        index=index,
+        limit=limit,
+        limit_direction=limit_direction,
+        limit_area=limit_area,
+        fill_value=fill_value,
+        **kwargs,
+    )
+
+    return _maybe_downcast(interp_values, downcast)
```
Comment (jbrockmendel) [inline pandas/core/missing.py:1044]: can this be done in the Manager method?

**Comment by jbrockmendel** [comment]
Can any of this be shared with the Block method?

**Comment by jorisvandenbossche** [comment]
> Can any of this be shared with the Block method?

I think the actual interpolate function is already shared (i.e. `interpolate_array_2d`). The rest is mostly some boilerplate that is slightly different (eg the Block method has an additional check to split the block). I think the `try/except` for `clean_fill_method` could be shared, but I am not sure it is worth creating another function for those few lines.

File: pandas/core/missing.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/core/missing.py:1044]: This is logic that belongs in the `interpolate` function? (for putting this in the manager, I would need to create a small wrapper function that combines the above method with `_maybe_downcast` to use in `ArrayManager.apply`. I don't really see an advantage of that)

Comment (jbrockmendel) [reply pandas/core/missing.py:1044]: Not a huge deal.  My thought is that post-op downcasting is likely to be the part where AM/BM are different, so putting the logic in the AM/BM method is more likely to be conducive to sharing.

Comment (jorisvandenbossche) [reply pandas/core/missing.py:1044]: I don't think that downcasting logic is different at the moment?

**Comment by jbrockmendel** [comment]
I increasingly think that this type of change should be left until right before BlockManager is ripped out. Until then, this just duplicates code and makes it easy for behavior to accidentally drift apart.

**Pull request closed without merging**